### ISSUE TITLE
setPropertyValue method in NamedElementImpl

### ISSUE THREAD

**Issue by lwrage** [lwrage/osate2-core#60]
<a href="https://github.com/mnam"><img src="https://avatars0.githubusercontent.com/u/1434565?v=4" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [mnam](https://github.com/mnam)**
_Mon May 21 04:23:17 2012_
_Originally opened as https://github.com/osate/osate2-core/issues/59_

----

I believe that there may be a bug here from changing to version 2

When ever we call this, it does a 
checkPropertyAssociation(pd, Collections.singletonList(pv));

and then in checkPropertyAssociation it has a 
if (!acceptsProperty(pd)) {
          throw new IllegalArgumentException("Property " + pd.getName() + " does not apply to " + getName());
}

when I use this I get caught by this exception when I'm not suppose to.
This happens when I am trying to associate a contained property to a classifier using this method.
And if the property does not apply to the classifier type, this throws an exception although
I am correctly using the property because it is associating the property to a subcomponent which
accepts the property.

Simple example is for a property type of 

Compute_Execution_Time: Time_Range applies to (thread, device, subprogram, event port, event data port);

Then if I try to assign this property to a thread that is a subcomponent of a process that is again a subcomponent
of a system by including in the the system implementation the following property

Compute_Execution_Time_Without_IO => 1 ms .. 2 ms applies to ProcessName.Thread_Name;

I can't save this in the system implementation by calling setPropertyValue for the system implementation 
because the system itself is not in the "applies to" of the property definition.

Of course, this property assignment is valid when typing in directly and should be.
When I include the system in the definition, it works. 

In conclusion, the setPropertyValue  method would cause wrongful exceptions when used with a Classifier and appliesTo. I think originally, it was only used for component instances which is why it was correct back then.

Should I just not use this function for my case? I could just skip the check.

Thanks.
Min Young

**Comment by lwrage** [comment]
<a href="https://github.com/juli1"><img src="https://avatars2.githubusercontent.com/u/993972?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [juli1](https://github.com/juli1)**
_Mon Dec 24 14:44:08 2012_

----

Hi Min-Young,

I try to identify the bug and make it reproducible. Can you please provide a way for reproducing this bug please ? A model and some code example (of debugger session) would be useful. Then, from what I have seen, when you try to associate the property, here, you try to associate it with the system. But on the other hand, the property must be associated with the thread and thus, you should try to call setPropertyValue to the thread instance itself.

But since this bug report is quite old, it may be oudated or also already fixed. Please let us know how if this is still occuring and how we can reproduce and fix it.

Thanks,

**Comment by lwrage** [comment]
<a href="https://github.com/mnam"><img src="https://avatars0.githubusercontent.com/u/1434565?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [mnam](https://github.com/mnam)**
_Mon Dec 24 18:10:14 2012_

----

Hi Julien,

Yeh, this is an old one but I believe hasn't been fixed. Unless it was done in the last 4 months.
I believe you have misunderstood some parts of my original posting. Rereading it, from a fresh mind
even I feel that it is complicated  than it needs to be.

So this setPropertyValue method has no problem when it is used for an instance.
It use to only exist for instances but now it is moved to the NamedElement.

Problem occurs when someone tries to use it for a ComponentClassifier.
For example, what I am doing is creating a property association in the following system component type.

system implementation A.impl
  subcomponents
    p1; process processType;  
  properties
    Compute_Execution_Time => 2ms applies to p1.threadName;  --want to create this in code.
end A.impl;

One would create a PropertyValue and call 

public final PropertyAssociation setPropertyValue(final Property pd, final PropertyValue pv) {
   this.checkPropertyAssociation(pd, Collections.singletonList(pv));
   ...
   this.getOwnedPropertyAssociations().add(pa);
   return pa;
}
on the system implementation to set the property value in the system implementation.

However, this is immediatly stopped by the exception in checkPropertyAssociation which basically
says that Compute_Execution_Time can't be assigned to system implementation A.impl 
although I'm giving it the property value to be applied to its subcomponent thread and not the system itself.

Hope this explains it.
It's not a problem when used on instances.

Thanks.
Sincerely,
Min Young

**Comment by lwrage** [comment]
<a href="https://github.com/juli1"><img src="https://avatars2.githubusercontent.com/u/993972?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [juli1](https://github.com/juli1)**
_Wed Jan  9 20:33:39 2013_

----

Hi,

There as a bunch of updates in the meantime, so, it might be worth to double-check. If you still have this issue, please submit a project snippet with some sample code so that we can reproduce the issue on our side.

Otherwise, we night close this ticket.

Thanks,

Best,

**Comment by lwrage** [comment]
<a href="https://github.com/mnam"><img src="https://avatars0.githubusercontent.com/u/1434565?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [mnam](https://github.com/mnam)**
_Thu Jan 17 16:51:24 2013_

----

Yeh lets close this. Thanks.
